GDExtension: Mark virtual function as `is_required` in `extension_api.json`

dsnopek opened the pull request:
This is a draft exploring how to implement https://github.com/godotengine/godot-proposals/discussions/9982

TODO to finish:

- ~~Find all the required virtual methods and swap in the `GDVIRTUAL*_REQUIRED()` macro~~
- ~~Replace all `GDVIRTUAL_REQUIRED_CALL()` with `GDVIRTUAL_CALL()`~~

dsnopek committed:
GDExtension: Mark virtual function as `is_required` in `extension_api.json`

Co-authored-by: Jovan Gerodetti <jovan.gerodetti@titannano.de>

TitanNano commented:
@dsnopek I migrated all the remaining instances of `GDVIRTUAL_REQUIRED_CALL` in e07dbec346bcdb86dd22a0562c23772110e7b34f and removed the macro in 0590f31c893d3c26c58ad0a155b45b0cee76e0a1. You are welcome to add these commits to your PR.

dsnopek commented:
@TitanNano Thanks! I added your changes.

After I've gone through and double-checked all the changes I'll take this out of draft.

TitanNano commented:
@dsnopek do you think this can make it into 4.4?

dsnopek commented:
This would be great to get into 4.4, especially if we can get PR https://github.com/godotengine/godot/pull/86079 too. I'll try to find some time soon to dig back into both PRs.

dsnopek commented:
I just rebased, updated for new required virtual methods that have been added, and _finally_ went back through everything and double-checked that everything that should be required is marked as required.

So, I'm taking this one of draft status now!

Bromeon reviewed (approved):
This looks great, thanks so much! I implemented the Rust part and while I haven't had the chance to test individual APIs yet, the `is_required` field is consistently present for all virtual methods :slightly_smiling_face:

Maybe @TitanNano could also have another look?

One question, do you think `is_override_required` would be clearer than just `is_required`? Or clear enough from the context?

akien-mga commented on core/object/object.h:
```diff
@@ -215,6 +215,7 @@ enum MethodFlags {
 	METHOD_FLAG_VARARG = 16,
 	METHOD_FLAG_STATIC = 32,
 	METHOD_FLAG_OBJECT_CORE = 64,
+	METHOD_FLAG_VIRTUAL_REQUIRED = 128,
```
@reduz Can you confirm whether it's fine to add a new method flag?

akien-mga replied:
Tried to ask three times but no luck, so let's merge and reduz will tell us if this was incorrect :)

akien-mga commented:
Thanks!
